src/call/mod.rs

```diff
@@ -108,6 +110,126 @@ impl RpcStatus {
     }
 }
 
+/// `MessageReader` is a zero-copy reader for the message payload.
+///
+/// To achieve zero-copy, use the BufRead API `fill_buf` and `consume`
+/// to operate the reader.
+pub struct MessageReader {
+    buf: *mut GrpcByteBuffer,
+    reader: GrpcByteBufferReader,
+    slice: Option<GrpcSlice>,
+    // Hack: apparently its lifetime depends on `self.slice`. However
+    // it's not going to be accessed by others directly, hence it's safe
+    // to mark it static here.
+    bytes: &'static [u8],
+    length: usize,
+}
+
+impl MessageReader {
+    /// Get the available bytes count of the reader.
+    #[inline]
+    pub fn pending_bytes_count(&self) -> usize {
+        self.length
+    }
+
+    pub fn read_directly(&mut self, buf: &mut [u8]) -> usize {
```

Comment:
This one does not need you to `.unwrap()`, since it's logically safe.

Reply:
I think the compiler is smart enough to optimize `.unwrap()`. Besides, it is confusing that there are both `read` and `read_directly` but they are just basically the same.

Reply:
`.unwrap()` means you're ignoring some errors. it's quite common practice to redirect error handling using `?`s or `map`s or `flat_map`s but if there's no error we'll never need to do those.

You're actually handling the never-present errors in #134 like
https://github.com/pingcap/grpc-rs/blob/8fff94905ca9bccc1875f7f5a9cae7d12aa17196/src/call/mod.rs#L149-L153
https://github.com/pingcap/grpc-rs/blob/8fff94905ca9bccc1875f7f5a9cae7d12aa17196/src/call/mod.rs#L122

These are all unnecessary. I believe these codes will get optimized by the compiler, but it's not readable comparing to this version:
https://github.com/pingcap/grpc-rs/blob/7372c787b1c5e98a9c34c9fcf3b8d4fd3a772b11/src/call/mod.rs#L164-L167

Reply:
Those codes are just general implementation of read_to_end, feel free to use `unwrap` to replace them.

If `Read` trait is not implemented, then I'm OK to remove the `Result` wrapper. However, since `Read` trait is implemented, I prefer to just keep one unify interface.

Reply:
Ok, ok, ok.